Sparse loop device provisioning with systemd units

gdemonet opened the pull request:
**Component**: salt, storage

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:

The previous approach was relying entirely on Salt (through our custom
`metalk8s_volumes` execution/state module) to manage provisioning and
cleanup of loop devices (on sparse files).
The issues arise when (re)booting a node:
- Salt minion has to execute a startup state, which may come in when
kube-apiserver is not (yet) available
- `losetup --find` calls, with the version available for CentOS 7, are
not atomic, so we may lose some devices if we have too many (re-running
the state manually is the current workaround)

**Summary**:

To fix these, we introduce two main changes:
- management of loop devices is delegated to systemd; using a unit
template file, we define a service per volume (passing the Volume's UUID
as a parameter), which will provision and attach the device on start,
and detach it on stop
- `losetup --find` invocations from these units is made sequential by
using an exclusive lockfile (this is not perfect, but avoids the need to
re-implement the command ourselves to include the fix from higher
versions of `losetup`)

Note that we cannot simply use `losetup --detach` in the unit template,
because sparse volumes may not always be discovered from the same path:
- either the volume is formatted, and we can find it in
/dev/disk/by-uuid/
- or it's kept raw, so we only have the UUID in the partition table, and
we can discover the device through /dev/disk/by-partuuid/

**Acceptance criteria**:

- CI keeps working (provisioning and cleanup of loop devices is tested)
- (manual) can reboot a node with many sparse loop volumes (10+) and have them all in `losetup -l` after

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2726

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

NicolasT commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -651,7 +651,8 @@ def _get_parts(self) -> Iterator[str]:
     Path('salt/metalk8s/volumes/prepared/installed.sls'),
-    Path('salt/metalk8s/volumes/provisioned/init.sls'),
+    Path('salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service'),
+    Path('salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up'),
```
I'd say `cleanup` but OK ;)

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/salt/minion/files/minion-99-metalk8s.conf.j2:
```diff
@@ -10,8 +10,3 @@ use_superseded:
 
 saltenv: {{ saltenv }}
-
-# Prepare volume at startup (required for loop devices persistency).
-startup_states: sls
-sls_list:
-  - /metalk8s/volumes/provisioned
```
:heart:

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=Setup MetalK8s sparse loop device %I
+RequiresMountsFor=/var/lib/metalk8s/storage/sparse
```
We could add an `AssertFileNotEmpty=/var/lib/metalk8s/storage/sparse/%i` here as well, see https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Conditions%20and%20Asserts (I've been reading too many `systemd` man-pages recently...)

gdemonet replied:
Added.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,13 @@
+[Service]
+Type=oneshot
+ExecStart=/sbin/losetup --find --partscan "/var/lib/metalk8s/storage/sparse/%i"
+ExecStop=/usr/local/bin/metalk8s-volume-clean-up "%i"
```
I really think this should go in `/usr/local/bin/libexec`, not `/usr/local/bin`. It's not some executable supposed to be regularly executed by a system user.

Also, `cleanup` :wink:

TeddyAndrieux replied:
> It's not some executable supposed to be regularly executed by a system user.

`/usr/local/sbin` then no ?

NicolasT replied:
Nope, `sbin` is for executables to be used by a superuser. This script is only ever to be executed by `systemd`, not by any regular/interactive user, hence `libexec`.

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=Setup MetalK8s sparse loop device %I
```
```suggestion
Description=Setup MetalK8s sparse loop volume %I
```

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+def find_device(volume_id):
+    path_by_uuid = "/dev/disk/by-uuid/{}".format(volume_id)
+    if os.path.exists(path_by_uuid):
+        return path_by_uuid
```
Could actually read the symlink here to always return `/dev/loopN` from this function.

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+        os.path.realpath("/sys/class/block/{}/..".format(partition_name))
+    )
+
+    return "/dev/{}".format(device_name)
```
Add an assertion to check this actually exists? It always should, but shouldn't harm...

gdemonet replied:
Done (and also checking that this is a loop device)

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+def clean_up(device_path):
+    try:
+        subprocess.check_output(
+            ["losetup", "-d", device_path],
```
The original implementation in `salt/_modules/metalk8s_volumes.py` did the `losetup -d` in pure Python code. It could make sense to do the same instead of shelling out.

`losetup -d` implementation basically boils down to https://github.com/karelzak/util-linux/blob/a61955ba1f15bf0a252a03a9abe54c1e02828c2e/lib/loopdev.c#L1518

NicolasT replied:
Note that, *iff* we need to have a wrapper for the loop device creation (locking...) as well, we could consider to (re-)implement this in pure Python as well.

gdemonet replied:
Let's go for pure Python! Makes me think I have forgotten to remove the `losetup -d` from the `clean_up` method, in our module.

gdemonet replied:
Done (and removed the corresponding code from our custom module).

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+        ).format(exc.cmd, exc.returncode, exc.output))
+
+
+if __name__ == '__main__':
```
```suggestion
def main():
```
and
```
if __name__ == '__main__':
    main()
```

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+    clean_up(device_path)
+
+    print("Volume '{}' was successfully cleaned up".format(volume_id))
+    sys.exit(0)
```
Not needed.

gdemonet replied:
Removed.

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -3,51 +3,78 @@ include:
+    - user: root
+    - group : root
+    - mode: 644
+    - makedirs: True
```
No. If `/etc/systemd/system` doesn't exist then something is *really* wrong on the system and we shouldn't ignore that.

gdemonet replied:
Removed.

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -3,51 +3,78 @@ include:
+
+Install clean-up script:
+  file.managed:
+    - name: /usr/local/bin/metalk8s-volume-clean-up
```
`/usr/local/libexec`

gdemonet replied:
Done.

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -3,51 +3,78 @@ include:
+    - user: root
+    - group : root
+    - mode: 755
+    - makedirs: True
```
Meh. This should really exist on the system already.

gdemonet replied:
Removed.

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -3,51 +3,78 @@ include:
-      - metalk8s_volumes: Prepare backing storage for {{ volume }}
+      - metalk8s_volumes: Prepare backing storage for {{ volume_name }}
+    - watch:
+      - file: Set up systemd template unit for sparse loop device provisioning
```
Shouldn't this be a `require` instead of a `watch`? If it needs to be a `watch`, why exactly? We definitely don't want to 'restart' the service if for whatever reason we update the service template, because restarting will likely not be possible (since the loop device is in use by some K8s container).

gdemonet replied:
I thought `losetup --find` was idempotent. Using `watch` was indeed to reload the service in case the template changed (e.g. fixing a bug somewhere in that chain), but maybe we shouldn't do this at all.

NicolasT replied:
It's not:
```
[root@containerd ~]# losetup -l
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
/dev/loop1         0      0         0  0 /root/disk
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
/dev/loop1         0      0         0  0 /root/disk
/dev/loop2         0      0         0  0 /root/disk
```

Furthermore, a restart of the service would imply an `losetup -d` of the existing volume (unit is stopped), which we don't want to happen.

gdemonet replied:
Used a `require` instead.

bert-e commented:
# Branches have diverged



This pull request's source branch `bugfix/2726-sparse-file-systemd` has diverged from
`development/2.7` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.7` into `bugfix/2726-sparse-file-systemd`
* Rebase `bugfix/2726-sparse-file-systemd` onto `origin/development/2.7`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,13 @@
+RemainAfterExit=yes
+
+[Install]
+WantedBy=multi-user.target
```
missing newline at the end of the file (as GitHub pointed out ^^)

gdemonet replied:
Done.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python
```
We should specified 2 or 3 (seeing the import future I guess it's 2, but better be explicit).

NicolasT replied:
And somewhere make sure it's installed on the system. Sure, we can assume it is (we're running `salt-minion` after all), but still...

gdemonet replied:
Sticking with Python 2 for now, not exactly sure where we should check if it's installed though. In the service template? Or through Salt?

NicolasT replied:
Salt. This is where maybe at some point these files shouldn't be dropped as-is by Salt, but packaged properly, which then allows to add a dependency on some Python version.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+
+def find_device(volume_id):
+    path_by_uuid = "/dev/disk/by-uuid/{}".format(volume_id)
+    if os.path.exists(path_by_uuid):
```
Could even use `os.path.islink` (or Pathlib, but since it's Python 2, not really).

NicolasT replied:
It's a link because that's how `udev` generally sets things up. However, it *could* (theoretically) be a proper block device. Why rule that out?

slaperche-scality replied:
In that case, better keep the `os.path.exists` yeah.

TeddyAndrieux replied:
In metalk8s-2.7 we will likely go with Salt Python3 so we could likely assume we run with Python3 (but since it's not merged yet we may need to stick with Python2)

gdemonet replied:
Kept it.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+        return path_by_uuid
+
+    path_by_partuuid = "/dev/disk/by-partuuid/{}".format(volume_id)
+    if not os.path.exists(path_by_partuuid):
```
ditto

gdemonet replied:
Kept `os.path.exists`, we are going to check what we find anyway.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-volume-clean-up:
```diff
@@ -0,0 +1,61 @@
+    if not os.path.exists(path_by_partuuid):
+        die("Volume '{}' backing device was not found".format(volume_id))
+
+    partition_name = os.path.basename(os.path.realpath(path_by_partuuid))
+    device_name = os.path.basename(
+        os.path.realpath("/sys/class/block/{}/..".format(partition_name))
+    )
```
I wonder if using regex to extract `loop[0-9]+` isn't simpler here, but maybe that way is more reliable.

NicolasT replied:
A regex would make assumptions about the naming, which is not guaranteed. Here all we assume is how sysfs is structured, which is kernel API (I believe...).

We *could* assert the final path we calculate is an actual block device (from `stat`), and a loop device (also from `stat`, though not sure how to figure out the major of `loop` except by parsing `/proc/devices`).

NicolasT replied:
(Easy using `awk`: `awk '$2=="loop" {print $1}' /proc/devices`, and likely `7` on all systems, not dynamically registered, but still, hard-coding isn't great).

gdemonet replied:
I do prefer the regex as well. It's not like we're trying to make this logic work for "any kind of device partition", we **know** it is supposed to be a loop device, so if we don't find `loop[0-9]+`, it's a good idea to bail out (if some other OS we support decides to break this behaviour, our code will explicitly tell us).

NicolasT replied:
I do agree we want to assert it's a loop-device. However, we should not do so based on some name, but based on the major of the block-device matching the major assigned to the `loop` block-device driver.

gdemonet replied:
Added a check to verify that:
- the device path exists
- it's a block device
- its major is the "loop" major (set to 7, hardcoded, sorry @NicolasT)

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,13 @@
+Type=oneshot
+ExecStart=/sbin/losetup --find --partscan "/var/lib/metalk8s/storage/sparse/%i"
+ExecStop=/usr/local/bin/metalk8s-volume-clean-up "%i"
+TimeoutSec=60
```
Actually not sure we want/need such timeout.

gdemonet replied:
OK, indeed it's almost an instantaneous operation.

gdemonet replied:
Removed.

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+
+
+SPARSE_FILES_DIR = '/var/lib/metalk8s/storage/sparse/'
+LOOP_MAJOR = 7
```
You could cite [Documentation/admin-guide/devices.txt](https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt#L191) from linux source tree as reference.

e.g.:
```python
# See Documentation/admin-guide/devices.txt in linux source tree
 LOOP_MAJOR = 7 
```

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+def is_loop_device(path):
+    device_stat = os.stat(path)
+    return stat.S_ISBLK(device_stat.st_mode) \
+        and (_major(device_stat.st_rdev) == LOOP_MAJOR)
```
Could use [`os.major(device_stat.st_rdev)`](https://docs.python.org/2.7/library/os.html#os.major) instead of coding your own `_major` :wink:

NicolasT replied:
Why not `os.major`?

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+        help='Cleanup MetalK8s sparse loop volumes',
+    )
+
+    parser.add_argument(
```
If this is required, why use a flag argument instead of a positional one?

gdemonet replied:
Sorry, remnants of the temporary implementation where I had an `action` (`provision` or `cleanup`) as positional. Anyway, will change.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+    )
+
+    parser.add_argument(
+        '-i', '--uid',
```
It's not `uid`, but `uuid` (though again, not sure this should be a flag).

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+    parser.add_argument(
+        '-i', '--uid',
+        required=True,
+        help='UID of the MetalK8s Volume object',
```
`UUID`

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+import sys
+
+
+SPARSE_FILES_DIR = '/var/lib/metalk8s/storage/sparse/'
```
Can drop the last `/` (see below).

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+
+def check_sparse_file(volume_id):
+    """Check if the sparse file exists for this volume ID."""
+    sparse_file_path = SPARSE_FILES_DIR + volume_id
```
`os.path.join` iso `+`

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+    finally:
+        fd.close()
+
+    print("Loop device for volume '{}' was successfully detached".format(
```
The `Device already freed` message would go to `stderr`. Is that consistent?

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+
+# Helpers {{{
+def die(message, exit_code=1):
+    print(message, file=sys.stderr)
```
Does `sys.stderr` need a `flush` before `sys.exit`?

slaperche-scality replied:
`stderr` is usually unbuffered, but [it's not that clear in Python](https://stackoverflow.com/a/22478875) so we may want to flush to be safe (especially since the behavior is different between python 2 and 3, as pointed [here](https://bugs.python.org/msg149408)).

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+      is labeled with the volume ID, hence discoverable under
+      /dev/disk/by-partuuid/
+    """
+    path_by_uuid = "/dev/disk/by-uuid/{}".format(volume_id)
```
`os.path.join`.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+    if os.path.exists(path_by_uuid):
+        device_path = os.path.realpath(path_by_uuid)
+    else:
+        path_by_partuuid = "/dev/disk/by-partuuid/{}".format(volume_id)
```
`os.path.join`

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+        device_path = os.path.realpath(path_by_uuid)
+    else:
+        path_by_partuuid = "/dev/disk/by-partuuid/{}".format(volume_id)
+        if not os.path.exists(path_by_partuuid):
```
Now if we'd add a third type of device (discovery), this would become one more indented block, and so on. I'd put all 'discovery' code (per case) in a separate function, then call them one by one until one returns a non-`None` result (or doesn't raise, whatever), then fail if none of them return a 'good' value: cleaner code, better scoped, less nesting and easier to test.

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+
+        partition_name = os.path.basename(os.path.realpath(path_by_partuuid))
+        device_name = os.path.basename(
+            os.path.realpath("/sys/class/block/{}/..".format(partition_name))
```
`os.path.join` and `os.path.pardir`

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume-cleanup:
```diff
@@ -0,0 +1,128 @@
+        device_name = os.path.basename(
+            os.path.realpath("/sys/class/block/{}/..".format(partition_name))
+        )
+        device_path = '/dev/{}'.format(match.group(0))
```
I may be missing something, but where does `match` come from? May need to add this file to the linted Python files in buildchain?

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,14 @@
+[Service]
+Type=oneshot
+ExecStart=/bin/flock --exclusive --wait 10 /var/lock/metalk8s-sparse-volume.lock \
+             -c "/sbin/losetup --find --partscan /var/lib/metalk8s/storage/sparse/%i"
```
Could not have `-c "..."`, and use the `sudo`-style invocation, no?

NicolasT commented on salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service:
```diff
@@ -0,0 +1,14 @@
+Type=oneshot
+ExecStart=/bin/flock --exclusive --wait 10 /var/lock/metalk8s-sparse-volume.lock \
+             -c "/sbin/losetup --find --partscan /var/lib/metalk8s/storage/sparse/%i"
+ExecStop=/usr/local/bin/libexec/metalk8s-sparse-volume-cleanup --uid "%i"
```
`/usr/local/bin/libexec` ain't right.

gdemonet replied:
Whoops.

NicolasT commented on salt/metalk8s/volumes/unprepared/init.sls:
```diff
@@ -1,19 +1,27 @@
+{%- if to_remove is not none %}
+  {%- if 'sparseLoopDevice' in to_remove.spec %}
+Disable systemd provisioning of loop device for {{ volume_name }}:
+  service.disabled:
```
Will this also 'stop' the service?

gdemonet replied:
Good catch. I didn't realize it wasn't stopped, because I left the cleanup logic in the module :confused:

NicolasT reviewed (changes requested):
Some issues

gdemonet committed:
salt: Provision sparse volumes with systemd

The previous approach was relying entirely on Salt (through our custom
`metalk8s_volumes` execution/state module) to manage provisioning and
cleanup of loop devices (on sparse files).
The issues arise when (re)booting a node:
- Salt minion has to execute a startup state, which may come in when
kube-apiserver is not (yet) available
- `losetup --find` calls, with the version available for CentOS 7, are
not atomic, so we may lose some devices if we have too many (re-running
the state manually is the current workaround)

To fix these, we introduce two main changes:
- management of loop devices is delegated to systemd; using a unit
template file, we define a service per volume (passing the Volume's UUID
as a parameter), which will provision and attach the device on start,
and detach it on stop
- `losetup --find` invocations from these units is made sequential by
using an exclusive lockfile (this is not perfect, but avoids the need to
re-implement the command ourselves to include the fix from losetup 2.25
and higher: see karelzak/util-linux@f7e21185)

Note that we cannot simply use `losetup --detach` in the unit template,
because sparse volumes may not always be discovered from the same path:
- either the volume is formatted, and we can find it in
/dev/disk/by-uuid/
- or it's kept raw, so we only have the UUID in the partition table, and
we can discover the device through /dev/disk/by-partuuid/

Fixes: #2726

gdemonet committed:
lint: Run pylint for sparse volume cleanup script

We adjust the `lint-python` tox environment to:
- handle Python 2 and Python 3 files (we only run mypy on the latter)
- always run both pylint and mypy, and report exit codes for both in
case of failure

See: #2726

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/sparse_volume_cleanup.py:
```diff
@@ -0,0 +1,158 @@
+import sys
+
+
+SPARSE_FILES_DIR = '/var/lib/metalk8s/storage/sparse'
+
+# https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt#L191
+LOOP_MAJOR = 7
+
+# https://github.com/torvalds/linux/blob/master/include/uapi/linux/loop.h#L110
+LOOP_CLR_FD = 0x4C01
```
Since those constants are always used by a single function, maybe it's better to declare them near their usage.

gdemonet replied:
Good point, I just store them at the top of my modules, out of habit :)

NicolasT replied:
Hah, bike-shedding: I like them on top, actually.

gdemonet replied:
Ah sorry, missed your comment @NicolasT so already did the change. I think both sides have good arguments, but the stakes are too low for us to have this debate. So I won't change it anymore :stuck_out_tongue_winking_eye:

slaperche-scality commented on salt/metalk8s/volumes/prepared/files/sparse_volume_cleanup.py:
```diff
@@ -0,0 +1,158 @@
+        )
+
+    if not is_loop_device(device_path):
+        die(
```
minor: instead of having a `die` function, maybe raising an exception (+ global try/catch around `main()`) would be clearer (more idiomatic at least).
This (global try/catch) would also handle the errors you forgot/ignored on purpose (failure of `os.stat` and `os.open` for instance).

Not sure it's worth it though, it's your call.

gdemonet replied:
It's a good suggestion, and not too much work, I'll handle it.

gdemonet replied:
Did something really basic, should do the trick :slightly_smiling_face:

slaperche-scality commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -651,7 +651,8 @@ def _get_parts(self) -> Iterator[str]:
     Path('salt/metalk8s/volumes/prepared/init.sls'),
     Path('salt/metalk8s/volumes/prepared/installed.sls'),
-    Path('salt/metalk8s/volumes/provisioned/init.sls'),
+    Path('salt/metalk8s/volumes/prepared/files/metalk8s-sparse-volume@.service'),
```
About commit message: really good one :+1: 

On

> to include the fix from higher versions of `losetup`

Maybe add [a link to the fix](https://github.com/karelzak/util-linux/commit/f7e21185087c30c5f13bf92009238e316f8c1cf4) and/or mention in which version it was fixed (v2.25 IIRC), maybe when we migrate to CentOS 8 we can get rid of this workaround.

gdemonet replied:
Done!

gdemonet committed:
changelog: Mention sparse loop provisioning bugfix

See: #2726

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/2726-sparse-file-systemd` and the
destination branch `development/2.7`.

Please resolve the conflict on **the feature branch** (`bugfix/2726-sparse-file-systemd`).

```bash
 $ git fetch
 $ git checkout origin/bugfix/2726-sparse-file-systemd
 $ git merge origin/development/2.7
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:bugfix/2726-sparse-file-systemd
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented on salt/metalk8s/volumes/prepared/installed.sls:
```diff
@@ -17,3 +17,9 @@ Install gdisk:
+
+# Needed by the sparse volume cleanup script
+Ensure Python 2 is available:
+  test.fail_without_changes:
```
Well, I would've used a `pkg_installed('python2')` or something along those lines here ;-) But sure, works for this particular case I guess.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye gdemonet.
